### PR TITLE
Fix/delete account collections

### PR DESCRIPTION
**Pull request by leilahamm**
This pull request introduces several improvements and refactors to the account deletion flow in the settings. The changes ensure proper sequential execution of operations and enhance test coverage for these functionalities. The key updates are as follows:

- Added PlaylistViewModel as a Parameter:
  - Incorporated PlaylistViewModel into the AccountScreen in the settings to handle the deletion of user-owned playlists during account deletion.
- Refactored Deletion Callback:
  - Extracted the account deletion logic into a dedicated, reusable function to improve code readability and maintainability.
- Sequential Deletion Flow:
  - Enhanced the deletion process to follow a sequential order:
   - Verify the user's password.
   - Delete user-owned data in the following order:
     - Profile
     - Playlists
     - Map user data
   - Finally, delete the user account from authentication.
- New verifyPassword Function:
  - Introduced a verifyPassword function in the FirebaseAuthViewModel to validate the user's password before proceeding with the account deletion.
- Updated Deletion Functions:
  - Refactored the following functions to ensure proper sequential execution and error handling:
     - deleteOwnedPlaylists in PlaylistViewModel
     - deleteMapUser in MapUsersViewModel
     - deleteProfile in ProfileViewModel
- Test Adaptations:
  - Updated and expanded test cases to cover the new sequential deletion flow and refactored functions.
  - Ensured robust validation of edge cases, such as incorrect passwords or failure of specific deletion steps.

### CODE REVIEW

File: app/src/main/java/com/epfl/beatlink/ui/profile/settings/Account.kt [hunk 1]
```diff
@@ -161,3 +160,63 @@ fun AccountScreen(
...
+        withContext(Dispatchers.Main) { onDeletionFailed("Incorrect password") }
+        return@launch
+      }
+
+      // Step 2: Delete Firestore data
+      val deleteProfileResult = profileViewModel.deleteProfile()
+      val deleteMapUserResult = mapUsersViewModel.deleteMapUser()
+      val deletePlaylistsResult = playlistViewModel.deleteOwnedPlaylists()
+
+      if (!deleteProfileResult || !deleteMapUserResult || !deletePlaylistsResult) {
```
Comment (mikhailperevoznyk) [inline app/src/main/java/com/epfl/beatlink/ui/profile/settings/Account.kt:205]: This condition returns on a failed deletion, but the failed deletion could happen because of just the data not existing. This could happen with mapUser data and Playists data. So all data that existed will be deleted but because of a fail of deletion on a data that doesn't exist it will trigger this if condition and will not let us delete the account properly. And we will stay on the app with all data related to the account deleted except the firebase authentication data. You should find a fix to this. I recommend changing the behavior of delete function in the various viewModels.

File: app/src/main/java/com/epfl/beatlink/viewmodel/map/user/MapUsersViewModel.kt [hunk 2]
```diff
@@ -94,13 +94,23 @@ open class MapUsersViewModel(private val repository: MapUserRepository) : ViewMo
   }
 
   /** Delete a user from the database. */
-  fun deleteMapUser() {
-    viewModelScope.launch {
+  fun deleteMapUser(): Boolean {
```
Comment (mikhailperevoznyk) [inline app/src/main/java/com/epfl/beatlink/viewmodel/map/user/MapUsersViewModel.kt:97]: Needs to treat the case where the data does not exist. Needs to not give a failure on non existence of the data.

File: app/src/main/java/com/epfl/beatlink/viewmodel/library/PlaylistViewModel.kt [hunk 3]
```diff
@@ -204,10 +204,22 @@ class PlaylistViewModel(
     ?: emptyList() // Return an empty list if no playlist is selected
   }
 
-  fun deleteOwnedPlaylists() {
-    repository.deleteOwnedPlaylists(
-        onSuccess = { Log.d("PlaylistViewModel", "All playlists deleted successfully") },
-        onFailure = { e -> Log.e("PlaylistViewModel", "Failed to delete playlists", e) })
+  fun deleteOwnedPlaylists(): Boolean {
```
Comment (mikhailperevoznyk) [inline app/src/main/java/com/epfl/beatlink/viewmodel/library/PlaylistViewModel.kt:207]: Needs to treat the case where the data does not exist. Needs to not give a failure on non existence of the data.

File: app/src/main/java/com/epfl/beatlink/viewmodel/map/user/MapUsersViewModel.kt [hunk 4]
```diff
@@ -94,13 +94,23 @@ open class MapUsersViewModel(private val repository: MapUserRepository) : ViewMo
...
+          onSuccess = {
+            _mapUser.value = null
+            _playbackState.value = null
+            result = true
+          },
+          onFailure = {
+            Log.e("MapUserViewModel", "Failed to delete map user")
+            result = false
+          })
+      result
```
Comment (mikhailperevoznyk) [inline app/src/main/java/com/epfl/beatlink/viewmodel/map/user/MapUsersViewModel.kt:110]: the return of the result variable is not needed, just return directly true or false onSuccess or onFailure

File: app/src/androidTest/java/com/epfl/beatlink/ui/profile/settings/DeleteAccountButtonTest.kt [hunk 5]
```diff
@@ -213,40 +185,81 @@ class DeleteAccountButtonTest {
...
+    // Ensure no repository actions were triggered
+    coVerify(exactly = 0) { authRepository.verifyPassword(any()) }
+    coVerify(exactly = 0) { authRepository.deleteAccount(any(), any(), any()) }
 
-    // Verify that no navigation occurs
+    // Verify no navigation occurred
     verify(exactly = 0) { navigationActions.navigateToAndClearAllBackStack(Screen.WELCOME) }
   }
 
+  //    @Test
```
Comment (mikhailperevoznyk) [inline app/src/androidTest/java/com/epfl/beatlink/ui/profile/settings/DeleteAccountButtonTest.kt:203]: if this test is not needed you should delete instead of commenting it

**Review by mikhailperevoznyk — CHANGES_REQUESTED** [review]
You need to change the code so that the deletion works with non existent data, and some code clean up.

File: app/src/main/java/com/epfl/beatlink/viewmodel/profile/ProfileViewModel.kt [hunk 6]
```diff
@@ -196,14 +196,19 @@ open class ProfileViewModel(
...
-  open fun deleteProfile() {
-    val userId = repository.getUserId() ?: return
-    viewModelScope.launch {
+  open suspend fun deleteProfile(): Boolean {
+    val userId = repository.getUserId() ?: return false
+    return try {
       if (repository.deleteProfile(userId)) {
         _profile.value = null
+        true
       } else {
```
Comment (mikhailperevoznyk) [inline app/src/main/java/com/epfl/beatlink/viewmodel/profile/ProfileViewModel.kt:205]: this expression is not used

File: app/src/main/java/com/epfl/beatlink/viewmodel/profile/ProfileViewModel.kt [hunk 7]
```diff
@@ -196,14 +196,19 @@ open class ProfileViewModel(
...
         Log.e("DELETE_PROFILE", "Error deleting profile")
+        false
```
Comment (mikhailperevoznyk) [inline app/src/main/java/com/epfl/beatlink/viewmodel/profile/ProfileViewModel.kt:207]: this expression is not used

File: app/src/main/java/com/epfl/beatlink/viewmodel/library/PlaylistViewModel.kt [hunk 8]
```diff
@@ -204,10 +204,22 @@ class PlaylistViewModel(
...
+    return try {
+      repository.deleteOwnedPlaylists(
+          onSuccess = {
+            Log.d("PlaylistViewModel", "All playlists deleted successfully")
+            true
```
Comment (mikhailperevoznyk) [inline app/src/main/java/com/epfl/beatlink/viewmodel/library/PlaylistViewModel.kt:212]: this expression is not used

File: app/src/main/java/com/epfl/beatlink/viewmodel/library/PlaylistViewModel.kt [hunk 9]
```diff
@@ -204,10 +204,22 @@ class PlaylistViewModel(
...
+          },
+          onFailure = { e ->
+            Log.e("PlaylistViewModel", "Failed to delete playlists", e)
+            false
+          })
+      true
```
Comment (mikhailperevoznyk) [inline app/src/main/java/com/epfl/beatlink/viewmodel/library/PlaylistViewModel.kt:218]: this expression is not used

**Review by mikhailperevoznyk — CHANGES_REQUESTED** [review]
here are some expressions that are not used, you can delete them

File: app/src/main/java/com/epfl/beatlink/viewmodel/library/PlaylistViewModel.kt [hunk 10]
```diff
@@ -204,10 +204,22 @@ class PlaylistViewModel(
...
+  fun deleteOwnedPlaylists(): Boolean {
+    return try {
+      repository.deleteOwnedPlaylists(
+          onSuccess = {
+            Log.d("PlaylistViewModel", "All playlists deleted successfully")
+            true
+          },
+          onFailure = { e ->
+            Log.e("PlaylistViewModel", "Failed to delete playlists", e)
+            false
```
Comment (mikhailperevoznyk) [inline app/src/main/java/com/epfl/beatlink/viewmodel/library/PlaylistViewModel.kt:216]: This expression is not used

**Review by mikhailperevoznyk — APPROVED** [review]
the issues and changes were resolved. The code looks ready to be merged now.
